File: test_drive/lib/widgets/hybrid_editor/core/transactions/update_block_type_operation.dart
import '../block_node.dart';
import '../document.dart';
import 'operation.dart';

/// Operation to update the type of a block
class UpdateBlockTypeOperation implements Operation {
  final int blockIndex;
  final BlockType oldType;
  final BlockType newType;
  final String? oldLanguage;
  final String? newLanguage;

  const UpdateBlockTypeOperation({
    required this.blockIndex,
    required this.oldType,
    required this.newType,
    this.oldLanguage,
    this.newLanguage,
  });

  @override
  OperationType get type => OperationType.updateBlock;

  @override
  Document apply(Document document) {
    final block = document[blockIndex];
    return document.updateBlock(
      blockIndex,
      block.copyWith(type: newType, language: newLanguage),
    );
  }

  @override
  Operation inverse(Document document) {
    return UpdateBlockTypeOperation(
      blockIndex: blockIndex,
      oldType: newType,
      newType: oldType,
      oldLanguage: newLanguage,
      newLanguage: oldLanguage,
    );
  }
}

File: src/components/editor/core/operations/UpdateBlockTypeOperation.ts
import { BlockType, copyBlock } from "../BlockNode";
import { type Document, updateBlock } from "../Document";
import { type Operation, OperationType } from "./Operation";

/// Operation to update the type of a block
export class UpdateBlockTypeOperation implements Operation {
	readonly type = OperationType.updateBlock;

	constructor(
		readonly blockIndex: number,
		readonly oldType: BlockType,
		readonly newType: BlockType,
		readonly oldLanguage?: string,
		readonly newLanguage?: string,
	) {}

	apply(document: Document): Document {
		const block = document.blocks[this.blockIndex];
		let attributes = { ...block.attributes };
		if (this.newType === BlockType.checkboxList) {
			attributes = { ...attributes, checked: !!block.attributes?.checked };
		}
		if (this.newLanguage !== undefined) {
			attributes = { ...attributes, language: this.newLanguage };
		}
		return updateBlock(
			document,
			this.blockIndex,
			copyBlock(block, { type: this.newType, attributes }),
		);
	}

	inverse(document: Document): Operation {
		return new UpdateBlockTypeOperation(
			this.blockIndex,
			this.newType,
			this.oldType,
			this.newLanguage,
			this.oldLanguage,
		);
	}
}
